feat(dates) HowTo work with dates

abirembaut opened the pull request:
- Also Fix widgets list for date picker

Covers [BS-16083](https://bonitasoft.atlassian.net/browse/BS-16083) and [BS-16575](https://bonitasoft.atlassian.net/browse/BS-16575)

Once it is merged, 7.4 and 7.5 branches should be rebased onto 7.3 or 7.3 should be merged in those branches...

abirembaut committed:
feat(dates) HowTo work with dates

Once it is merged, 7.4 and 7.5 branches should be rebased onto 7.3 or 7.3 should be merged in those branches...

- Add images
- Update taxonomy

Covers [BS-16083](https://bonitasoft.atlassian.net/browse/BS-16083)

abirembaut commented:
WIP (Missing the part on how to display a date from the BDM)

nchabanoles commented on md/dates-management-tutorial.md:
```diff
@@ -0,0 +1,67 @@
+
+![BDM](images/dates-management-tutorial/tuto-dates-bdm.png) <!--{.img-responsive .img-thumbnail}-->
+
+Then, in the process, create a new business variable of the type of you business object.
```
Typo: you business object ==> your business object

nchabanoles commented on md/dates-management-tutorial.md:
```diff
@@ -0,0 +1,67 @@
+# How to work with dates in forms and BDM
```
Question: Do we plan to have a dedicated HowTo for BDM & Date manipulation in the process?
For instance when I want to use a Date attribute from a business object as input of a connector; in a condition for a transition; in the dynamic display name of a task...

abirembaut replied:
I don't think an additional tutorial is needed, but maybe a section in this one can be added.

abirembaut replied:
I will add a note (in case a string is used in the business) on how to convert it to a date for connectors expecting a java.util.Date

nchabanoles commented on md/dates-management-tutorial.md:
```diff
@@ -0,0 +1,67 @@
+
+Start by defining your business data model.
+
+It is recommended that you define the date attributes of your business data Objects with a **LONG** type instead of DATE if you want the value stored in the database to reflect the date entered by the user. Otherwise the timezone of the server will be used when converting the java date in an SQL date and, unless your server is configured in UTC, an offset will be applied. This will not be an issu ...
```
Do we really recommend Long to store a Date Only value?
What are we going to provide by default in 7.5+?

abirembaut replied:
We don't know yet for 7.5. But I can try with String to see how it goes.
The advantage of long compared to String is that there is no need to provide a validator in the contract to make sure that the date really exists. But this is true that if we move to String in 7.5, it may not be a good idea to advise Long...
I will try to implement the same process with Strings to see if it is doable before 7.5

nchabanoles commented on md/dates-management-tutorial.md:
```diff
@@ -0,0 +1,67 @@
+
+Start by defining your business data model.
+
+It is recommended that you define the date attributes of your business data Objects with a **LONG** type instead of DATE if you want the value stored in the database to reflect the date entered by the user. Otherwise the timezone of the server will be used when converting the java date in an SQL date and, unless your server is configured in UTC, an offset will be applied. This will not be an issu ...
```
Naively I would have explained the UC with Date in every layer and explained the constraints (such as the timezone shift operated automatically).
If for some reason the constraint is not acceptable, then the recommendation would be to use another type than Date. It would seem more didactic this way: explain how things work and better understand the impacts...

Here I can hear the question that we do not answer: "then what is the type Date used for in BDM"?

Or maybe it means that we should also present another use case that leverage the type date (a UC with Date and Time). What do you think?

abirembaut replied:
Actually both UC are explained in the rest of the tutorial (with Date and with Long).
Bug I understand the first point. I will try to present it the other way arround : How it works with dates and then use Long (or even String if it is possible (need to test it)) if the constraint is not acceptable.

nchabanoles commented on md/dates-management-tutorial.md:
```diff
@@ -0,0 +1,67 @@
+## Define the contract and task operations or business variable initialization script
+
+Now you need to define the *process instantiation* or *task execution* contract (depending on your use case).  
+The date picker widget output is an ISO 8601 formatted String ("2017-01-13T00:00:00.000Z") that will be automatically formatted to a java.util.Date by process instantiation or task execution REST API if the contract input is of type **DATE**. The date/time received will be considered as UTC. The time is set to midnight by the widget as it doesn't handle time.  
```
"will be considered as UTC" or "must be expressed in UTC" ?

abirembaut replied:
"Will be considered" as the contract supports formats that do not include the "z" like "2017-01-13T00:00:00". So you can pass a date that is not a UTC date, but it will be considered as UTC when it is converted to a Date object.
Maybe I should say "will be considered as UTC when it is converted to a Date object"

abirembaut committed:
feat(dates) HowTo work with dates

- Encourage String usage instead of Long
- Fix date widget information as its value is a Date object and not a
String
- Also Fix widgets list for date time picker
- shorten ToC with ### level titles set to #### level

Covers [BS-16083](https://bonitasoft.atlassian.net/browse/BS-16083)

pooravbos commented on md/dates-management-tutorial.md:
```diff
@@ -0,0 +1,103 @@
+![Variable initialization](images/dates-management-tutorial/tuto-dates-variable-init.png) <!--{.img-responsive .img-thumbnail}-->
+
+::: info
+If you want you can trunkate the ISO 8601 formatted String received from the contract input to only keep the date and drop the time information that is not relevant.
```
+If you want you can truncate the ISO 8601 formatted String received from the contract input to only keep the date and drop the time information that is not relevant.

abirembaut replied:
fixed

pooravbos reviewed (approved):
Updating a small typo

abirembaut commented:
Fixed API URL expression
link to the howto in the widets page

abirembaut commented:
retest this please

NathalieC commented on md/dates-management-tutorial.md:
```diff
@@ -0,0 +1,221 @@
+
+1. In the studio, open the Business Data Model editor (in menu **Development** > **Business Data Model** > **Manage...**)
+1. Create a **business object** *Employee* with an attribute *birthdate* of type **STRING**  
+![BDM](images/dates-management-tutorial/tuto-dates-bdm.png) <!--{.img-responsive .img-thumbnail}-->  
```
the BDM shows more attributes than expected.
May be worth adjusting

NathalieC commented on md/dates-management-tutorial.md:
```diff
@@ -0,0 +1,221 @@
+1. Select the business variable *employee*
+1. Follow the steps to generate a contract with a COMPLEX input *employeeInput* containing a *birthdate* attribute of type **DATE** (keep the option to generate the variable initialization script automatically)
+
+![Contract](images/dates-management-tutorial/tuto-dates-contract-date.png) <!--{.img-responsive .img-thumbnail}-->
```
the contract shows more attributes than expected from the business data chapter.
May be worth adjusting

NathalieC commented on md/dates-management-tutorial.md:
```diff
@@ -0,0 +1,221 @@
+1. Select the business variable *employee*
+1. Follow the steps to generate a contract with a COMPLEX input *employeeInput* containing a *birthdate* attribute of type **TEXT** (keep the option to generate the variable initialization script automatically)
+
+![Contract](images/dates-management-tutorial/tuto-dates-contract-text.png) <!--{.img-responsive .img-thumbnail}-->
```
the contract shows more attributes than expected from the business data chapter.
May be worth adjusting

NathalieC commented on md/dates-management-tutorial.md:
```diff
@@ -0,0 +1,221 @@
+1. Since you chose the type **TEXT** for the date contract input, a text **Input** widget will be added to the page for the contract input instead of a **Date picker** widget.  
+You will need to add and configure the **Date picker** widget yourself in order to replace the text input widget:  
+   1. From the **Form editor** palette, drop a **Date picker** widget on the whiteboard
+   1. Set the **Date picker** widget **Value** property to the same **Value** as the text **Input** widget
```
formInput.employeeInput.birthDate

NathalieC commented on md/dates-management-tutorial.md:
```diff
@@ -0,0 +1,221 @@
+You will need to add and configure the **Date picker** widget yourself in order to replace the text input widget:  
+   1. From the **Form editor** palette, drop a **Date picker** widget on the whiteboard
+   1. Set the **Date picker** widget **Value** property to the same **Value** as the text **Input** widget
+   1. Set the **Date picker** widget **Label** property to the same **Label** as the text **Input** widget
```
Birth Date

NathalieC commented on md/dates-management-tutorial.md:
```diff
@@ -0,0 +1,221 @@
+1. In the UI Designer, define a variable to retrieve the *Employee* currently stored using the context:  
+   1. In the **Variables** panel at the bottom of the editor, click on **Create a new variable**
+   1. Enter the name *employee* for the new variable
+   1. select the **type** *external API*
```
Select

abirembaut commented on md/dates-management-tutorial.md:
```diff
@@ -7,54 +7,54 @@ This howto applies to Bonita BPM versions prior to 7.5.0
+The following sections show how to create, update, and display an *Employee* object with a birth date stored either as a DATE or a STRING in the BDM.  
 
-## With a DATE attribute
+## Business object attributes with type DATE
```
Maybe
"Business object with DATE attributes"
or
"Business object attributes of type DATE"
?

abirembaut commented on md/dates-management-tutorial.md:
```diff
@@ -7,54 +7,54 @@ This howto applies to Bonita BPM versions prior to 7.5.0
-Otherwise the time zone of the server will be used when converting the java Date in an SQL **TIMESTAMP** and, unless your server is configured in UTC, an offset will be applied.  
-This will not be an issue if you do not query the business data database directly and only use the API (as the same conversion will be performed the other way around when retrieving the value from the database). In this case, you can safely use a **DATE** attribute type.
+If you only use the API to retrieve dates, and you do not need to query the business data database directly, you can safely use the type **DATE** for your business objects.
+If you query the database directly, use the type **STRING** instead of **DATE**. Otherwise the time zone of the server will be used when converting the java Date in an SQL **TIMESTAMP** and, unless your server is configured in UTC, an offset will be applied. Therefore, the value stored in the database will not reflect the date entered by the user.
```
"If you **need to** query the database directly" would seem more accurate to me.
Maybe you don't query the DB right now, be you will need to at one point. And it is consistent with the previous sentence "and you do not need to query the business data database"

abirembaut reviewed (commented):
LGTM, just 2 small comment
